### PR TITLE
feat: typography grid refinements

### PR DESCRIPTION
**Pull request by zchsh**
## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

This PR makes some refinements to the grid used for content in our "sidebar sidecar" layout. Specifically:

### Main content area tweaks

- Adds a `mainWidth` prop to `SidebarSidecarLayout`
    - This allows us to set a narrower max-width for long-form content pages
- For `mainWidth === 'narrow'`, tweaks the main content area `max-width`:
    - The main area max-width was previously `896px`, is now `680px` for `narrow` layouts
- On small viewports, now enforces `max-width`
    - This is especially relevant for `narrow` views
- Retains centering of main content area across viewports
- Uses `mainWidth="narrow"` for long-form docs and tutorials pages

### Sidecar tweaks

- Tweaks sidecar width
   - For `wide` main areas, retains `220px` sidecar width
   - For `narrow` main areas, increases sidecar width to`288px`
- Tweaks right padding to viewport from sidecar: was `48px`, is now `24px`
- Tweaks content-sidecar spacing: was `48px`, still `48px`, but is now `64px` for `1440px` and larger viewports

### Sidebar changes

- Tweaks padding in sidebar: was `24px`, is now `16px`
- Removes drop shadow on sidebar: was higher surface, now "surface base"

### Version switcher overlap fix

- Tweaks version switcher overlap logic for `<h1 />`
   - The version switcher is ~ `160px` wide. We used a margin to account for this width.
   - The margin was `25%`, which worked on large viewports, but broke down on smaller viewports
   - The margin is now `180px`
   - This is an imperfect fix, but didn't want to blow up scope here and get too far into a layout refactor. I think ideally we'd use a `flex` layout, that holds both the `h1` and version selector, so that we're not guessing at this width.

## 📸 Design Screenshots

> **Note**: screenshots are taken at 1920px viewport widths in order to clearly show the max-width differences.

| View | Main Area Width | Before | After |
| - | - | - | - |
| Tutorial library <br /> [/tutorials/library][/tutorials/library] | `wide` | tutorial-lib-before | tutorial-lib-after |
| Product landing <br /> [/waypoint][/waypoint] | `wide` | product-landing-before | product-landing-after |
| Product Tutorial landing <br /> [/waypoint/tutorials][/waypoint/tutorials] | `wide` | product-tutorials-landing-before | product-tutorials-landing-after |
| Collection page <br /> [/waypoint/tutorials/get-started-docker][/waypoint/tutorials/get-started-docker] | `wide` | collection-before | collection-after |
| Docs landing <br /> [/waypoint/docs][/waypoint/docs] | `wide` | docs-landing-before | docs-landing-after |
| Downloads <br /> [/waypoint/downloads][/waypoint/downloads] | `wide` | downloads-before | downloads-after |
| Docs page <br /> [/waypoint/docs/getting-started][/waypoint/docs/getting-started] | `narrow` | docs-before | docs-after |
| Tutorial page <br /> [/waypoint/tutorials/get-started-docker/get-started-docker][/waypoint/tutorials/get-started-docker/get-started-docker] | `narrow` | tutorials-before | tutorials-after |

## 🧪 Testing

- [ ] Visit pages that use the "sidebar sidecar" layout with the new "narrow" width
    - The content area should be narrower than upstream (680 px)
    - Pages are listed in the table above
- [ ] Visit pages that use the "sidebar sidecar" layout with the "wide" width, similar to current layout
    - The content area should be similar to upstream (896 px)
    - Pages are listed in the table above
- [ ] Visit pages that do not use "sidebar sidecar" layout
    - These pages should be unaffected, and identical to upstream
    - Examples: [home][preview], [/certifications][/certifications], 

## 💭 Anything else?

Not at the moment!

[preview]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/
[/certifications]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/certifications
[task]: https://app.asana.com/0/0/1203453209159722/f
[/tutorials/library]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/tutorials/library
[/vault/api-docs/secret/ad]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/vault/api-docs/secret/ad
[/terraform/docs]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/terraform/docs
[/waypoint]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/waypoint
[/waypoint/tutorials]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/waypoint/tutorials
[/waypoint/tutorials/get-started-docker]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/waypoint/tutorials/get-started-docker
[/waypoint/tutorials/get-started-docker/get-started-docker]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/waypoint/tutorials/get-started-docker/get-started-docker
[/waypoint/docs]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/waypoint/docs
[/waypoint/docs/getting-started]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/waypoint/docs/getting-started
[/waypoint/downloads]: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/waypoint/downloads

### CODE REVIEW

**Comment by zchsh** [comment]
@braydenlove This PR implements the "sidebar sidecar" layout grid changes from Figma.

Note that it's not exactly 1:1, since some parts of the existing implementation aren't 1:1. Most glaringly, we switch to our mobile menu lose the sidebar at `1200px` viewports, rather than `949px` as designed. Note the loss of the sidebar and the switch to the mobile menu are strictly tied in current designs, so we have to make both switches at the same time based on available content space in the top nav.

With this in mind, I've left the main content area centered across all viewports, rather than left aligned when the sidebar first disappears. Otherwise, with the sidebar disappearing so early, at `1200px` when the sidebar first disappears we'd get something like this:

Screen Shot 2022-12-05 at 18 34 42

Let me know if I'm jumping the gun there, happy to switch back to left-aligning as above if that's the preference 🙇

**Comment by braydenlove** [comment]
@zchsh thanks for spinning this up! We do need to find a way to switch to mobile nav later as we are doing it very early (in the smaller laptop zone) and are getting feedback from users about the awk early switch in regards to finding content easily. Do you have any ideas? I am seeing that a lot of sites are switching to a ... icon with a dropdown when zoomed. Is truncation an option?

**Comment by braydenlove** [comment]
Or perhaps we could add a scroll to the top bar nav bar?

**Comment by zchsh** [comment]
@braydenlove Re: revising the top nav to switch to mobile earlier, I think this is likely something we should approach separately from the typography grid work (either before or after).

If we want to address the top nav mobile behaviour _before_ the typography work here in the PR, I think we'll need to defer this PR and rethink prioritization. If we're okay with addressing it _after_ the typography work, then maybe there are incremental changes we could make here to mitigate things?

My hot take is the revised typography grid work is much better even with the mobile nav issues still in place... so might be worth focusing on typography work and later coming back to the nav.

## Re: truncation

Truncation might be an an option - though for Dev Dot I think switching to truncation is a relatively high-complexity solution to reach for, and I personally think showing some links outside a `...` and some within is less clear than showing either all links exposed, or all links within a menu dropdown or similar.

If we do want to move forward with truncation, we'd need some details on what behaviour we're aiming for. My assumption would be that we'd take the right-most links in the top nav, and try to detect overflow, and if there's overflow then put right-most links into a `...` dropdown until there's enough space available to show both the remaining links and the `...` dropdown. However, I don't think this would be a fully robust solution -  I think we'd run into additional complexity if the item being put in the `...` is a dropdown itself - I think we'd need to handle the case where `Documentation ↓` for example gets collapsed into the `...` truncated dropdown (this could happen on text zoom, for example, but could also happen if top nav items are re-ordered). I think robust behaviour rules could be a bit tough to figure out.

Another option might be removing eg the `Home` link - that might give us more space to tweak the mobile breakpoint. Maybe there's other content-focused stuff we could do in the top nav as well, such as collapsing other items into dropdowns.

## Re: other ideas

TLDR; I think revisiting two-tiered nav concepts would be worthwhile.

I think the truncation & content-tweak type approaches could get us somewhere, but I think a broader top nav refactor might get us even farther, and address challenges such as navigation on non-product areas of the site (such as Certifications).

One thing that sticks out to me with this context in mind is the role of links in navigation at a high-level:
1. **Site-wide links**, such as switching between products, Search, and auth dropdown
2. **Product-specific links**, such as switching between Tutorials, Docs, Install, and other product-specific outbound links (eg "Try cloud")
3. **Sidebar** for pages like Tutorials and Docs (on some other pages may be worth keeping as well, but in some cases it's duplicative)

We have other areas where we've approached somewhat analogous challenges, such as hashicorp.com - see for example the two-tiered nav and mobile behaviour on <https://www.hashicorp.com/products/vault>:

CleanShot 2022-12-06 at 13 28 44@2x

The approach there has two significant benefits IMO:
- Separate nav tiers create more horizontal room, allowing the full navs to show at smaller viewports, even as more links are added
- Separate mobile nav panes for the "site-wide" vs "product-specific" contexts mean navs can switch to "mobile" independently. Specifically at ~ 1150px wide, you can see the "product-specific" links are not yet in "mobile" mode, while the "site-wide" links are.

The hashicorp.com use case definitely differs in other ways, but some of the specifics around the two-tiered approach, particularly on mobile, seemed relevant!

Also worth noting that on dot-io sites we took a similar approach, except the "second tier" was the sidebar nav:

CleanShot 2022-12-06 at 13 28 48@2x

I wonder whether there's a revised approach to mobile behaviour for the sidebar nav, such as an independent pill or something to open up the sidebar navigation independently (as in mobile mockup below) rather than roll it into the "product-level" or "site-level" nav (which are currently all rolled into one). I think this type of approach could provide more consistent & straightforward access to all 3 layers of the nav across viewports, and could give us flexibility going forward in tweaking mobile behaviour at each of these layers:

CleanShot 2022-12-06 at 13 28 52@2x

Long wall of text there, but hopefully that's helpful 🙇

**Comment by zchsh** [comment]
@braydenlove Have heard this idea discussed so it's probably not new, but dropping the `Home` link from the top menu would be pretty quick - it'd be a one line change, removing [this line of code](https://github.com/hashicorp/dev-portal/blob/55d2c0293ca0a4c63232c3fc6ec8d500f9577490/src/components/navigation-header/components/product-page-content/utils/get-nav-items.ts#L61). I'll open a PR to demo it. That could be a quick way to start shaving down the mobile nav breakpoint.

With the context of this PR in mind, @kendallstrautman @braydenlove are there are revisions here we want to make to the typography grid work here? While I think we could potentially put this PR on hold while we work on other facets related to typography, I think it'd be preferable to try to get this work finalized in relative isolation, if that makes sense.

**Comment by braydenlove** [comment]
@zchsh yeah I think removing home is an ez win. I already created a ticket for it [here](https://app.asana.com/0/1203461693655903/1203452654466343/f), we will add more stuff as we get aligned. That being said I agree that this IA conversation is separate from typography and I will like to focus and get this out before we jump into the text zoom nav and ia quick wins work if that is cool with everyone else.

For revisions I would like us to increase the width of the sidecar to match the designs. Also I am unclear if we can push this stuff live like we did with the inline links and code-200... I will get with Jasmine to discuss. In the meantime do whatever you can (plz protect your time and focus on certifications), you now have the refined type stack, spacing and grids to play around with. Once we get all that in a single PR (ideally) we can use that as a prototype for the [RFC](https://docs.google.com/document/d/1POtogMsY-bFH4DQwFEq6ulWBZMrhs2uqahVZfGHZYws/edit)?

**Comment by zchsh** [comment]
@braydenlove Sounds good, very on board with focusing on the typography grid work in this PR 👍

## Re: sidecar width

I'd thought we were ignoring the sidecar for now, until revised designs are finalized. I think the width I can pull from current mockups might not be exact, since it gets really close to the edge of the page in the mockups, where on the live site it'll eat up more of the content area (shown around the smallest viewport where sidecar is present):

| Design | This PR |
| - | - |
| CleanShot 2022-12-07 at 13 35 41@2x | CleanShot 2022-12-07 at 13 42 13@2x |





We could adjust the "hide sidecar" breakpoint if we want there, or come up with a revised pixel value (currently `220px`, mockup shows `296px` I think?)

## Re: refined type stack, spacing, and grids

I've opened up https://github.com/hashicorp/dev-portal/pull/1419 for experimentation. Still need to add in the type stack there, but controls to play around with the spacing and content area max-width should be there. Will continue the discussion in that PR.

**Comment by braydenlove** [comment]
@zchsh curious what you think about [this](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/Developer?node-id=16612%3A246826&t=6jvgpOS8dzGMpjOP-1). Essentially make the max width of the sidecar (keep designs as they are today just increase width) to `312px` (with `24px` right margin) or 288px however you want to do it.

```
.sidebar-sidecar-layout_root {
--sidecar-width: 288px;
}
```

I have the gap between the sidebar/content-area/sidecar at 64px for bigger breakpoints then drop to 48px as we go down. LMK if that is too complex or if you have other ideas! We can also chat about this when meet later this week if easier.

**Comment by heitzke** [comment]
Agreed with all the above comments of separation of concerns here. We DEFINITELY need to address some items with the navigation but it's apparent that it'll be a bigger effort alone. The typography changes are looking good though and appreciate all the partnership here 👍

**Comment by zchsh** [comment]
@braydenlove I dig the increased size of the sidecar 👍 Patched up a related visual bug with the version select, and I think things are good to go. Summary of changes:

- Tweaks sidecar width: was `220px`, is now `288px`
   - Also tweaks right padding to viewport from sidecar: was `48px`, is now `24px`
   - Also tweaks content-sidecar spacing: was `48px`, still `48px`, but is now `64px` for `1440px` and larger viewports

Previews below are at the smallest viewports where the sidecar is shown.

## Before (live site currently)

before

## After (this PR)

after

## Note on content-sidecar spacing

With the current right-aligned sidecar, the spacing will appear larger than 48px in many cases due to whitespace. I think this is fine, and will be refined once we have the revised sidecar implemented, but wanted to make a note of it for clarity:

note-on-spacing

Personally I think this is looking great! 🎉  Happy to revise further if there's other stuff we want to tweak - though if we're overall happy here, I think we could also merge this in and tweak further in subsequent PRs 👍

**Review by braydenlove — COMMENTED** [review]
This is look really great thank you! Curious why pages like [this](https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/packer/plugins), [this](https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/packer/plugins/builders/oneandone), [this](https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/terraform/cloud-docs) aren't respecting the max 680px? Are we considering those landing pages? If so is there an easy way to make sure they do respect the max-width?

**Comment by zchsh** [comment]
@braydenlove Ah, great catches!

Re: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/terraform/cloud-docs, this is treated as a "landing page" similar to https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/terraform/docs. We could apply the "narrow" 680px width to all docs pages, even if they're landing pages 🤔 Or we could apply the "narrow" 680px to all landing pages EXCEPT `/<product>/docs` pages. Let me know what you think, in theory anything should be possible if we can nail down a clear heuristic for which pages we want the "narrow" 680px width to apply to 🙇 

Re: https://dev-portal-git-zstypography-grid-refinement-hashicorp.vercel.app/packer/plugins, this is a slip-up on my end - the Packer plugin pages are a separate layout that I need to fix up 😅 Will do that now, will be a quick fix.

**Comment by zchsh** [comment]
@braydenlove Updates based on that last comment:

- Updated so that only `/<product>/docs` pages with use the "wide" 896px max-width. All other pages (eg `/terraform/cloud-docs`) will use the "narrow" 680px max-width (even if they're the "root" of a base docs path, like `cloud-docs` in this case)
- Fixed the issue with Packer plugins, all `/packer/plugins` pages should now have the "narrow" 680px width

Changes will take a few minutes to deploy, but preview should be updated soon!

**Comment by braydenlove** [comment]
> 

Awesome man no worries at all, yeah I am wondering if we just respect the 680 max everywhere besides `/<product>`, `/<product>/downloads`, `/<product>/docs`, homepage, `/certifications`, `/tutorials/library`... Thoughts?

**Comment by zchsh** [comment]
@braydenlove Sweet, I think we're on the same page!

I've been thinking of it in the inverse a bit:
- We use the `wide` 896px max-width by default
- For specific views (listed below) we opt-in to the `narrow` 680px max-width.
    - All individual tutorial views (only tutorial content, not `/<product>/tutorials` or collection pages)
    - All individual docs views (_except_ `/<product>/docs`, where we render our custom cards-y layout)

Let me know if that lines up 🙇

**Comment by braydenlove** [comment]
> 

Yup I think we are aligned, I agree thinking about it inversely makes more sense. Thanks man!

**Comment by zchsh** [comment]
@braydenlove I think this PR is ready for final review - TLDR adjusts the grid for docs and learn pages, but see the PR description for details. Let me know what you think 🙇

**Comment by braydenlove** [comment]
@zchsh looks beautiful, just reviewed with @heitzke @kellyschummer and we all give the 👍🏼 up! 🚢 it

**Review by kendallstrautman — APPROVED** [review]
LGTM! :shipit:
